protos/idl/service.proto

```diff
@@ -98,3 +109,68 @@ message Tag {
 message Metadata {
     map<string, string> key_map = 1; // key map is a dictionary of key/val strings that represent metadata
 }
+
+message FilterExpression {
+    repeated SingleValueFilter filters = 1;
+}
+
+message SingleValueFilter {
+    oneof entity {
+        TagFilter tagFilter = 1;
+        PartitionFilter partition_filter = 2;
+        ArtifactFilter artifact_filter = 3;
+        DatasetFilter dataset_filter = 4;
+    }
+
+    // as use-cases come up we can add more operators, ex: gte, like, not eq etc.
+    enum ComparisonOperator {
+        EQUALS = 0;
+    }
+
+    ComparisonOperator operator = 10; // field 10 in case we add more entities to query
+    // Next field number: 11
+}
+
+message ArtifactFilter {
+    oneof property {
+        string artifact_id = 1;
+    }
+}
+
+message TagFilter {
+    oneof property {
+        string tag_name = 1;
+    }
+}
+
+message PartitionFilter {
+    oneof property {
+        KeyValuePair key_val = 1;
```

Comment:
Why is this a `oneof`? What other stuff could we put in this message in the future?

Reply:
Yea good question. This is for extensibility. For example, later we may add `Partition.propertyX`. We want to be able to add `propertyX` as something we can filter on.

Reply:
Sure, that sounds good.